docker-compose.yml

```diff
@@ -20,6 +20,10 @@ services:
       - 6379
     command: ["redis-server", "--save", "''", "--appendonly", "no"]
   mongo:
+    deploy:
```

Comment:
You'd be changing this for all users of the file. Admittedly, this file should not just be used off the shelf as-is, only people with docker/compose experience should use it, and should customize it to their needs. I have used this, and still do, but on machines with much more memory. I understand your use case (and others) would benefit for this, but what about those with different use cases? Should this be in a comment as a suggested tweak for low memory cloud users? Should be be in a separate compose with other tweaks for this use case? etc? I won't take this as-is, but I'm open to other solutions/alternatives!

Reply:
I think there are a couple of issues with the current configuration, let me lay those out:

1. The default configuration is not respecting standard operation practices of mongodb. Mongo db's default is to assume that it has exclusive access to all the memory in the system and that there are no other services running. The docker compose setup of cryptostore co-locates mongodb together with redis and the cryptostore / cryptofeed processes on the same host. For such a scenario limiting the memory of mongodb is strongly recommended due to mongodb's aggressive caching. https://docs.mongodb.com/manual/administration/production-notes/#id4

2. The default configuration does not recommend a hardware configuration. I didnt manage to get the docker compose for the default configuration to survive on less than 16GB of main memory. This is very confusing especially considering the following approximation of the workload:
`3 (pairs) * 2 (bid/ask) * 10000 (depth) * 8 (precision) * 5 (2xts/amt/price/booleans) = 2.4MB`

This is a reasonable lower bound on the memory needed to store all order books of the default configuration. Assuming less than 20000 orders between snapshots storing the deltas on top of a full snapshot should only increase the memory usage by factor 2. As crypto store is currently 1000x above the theoretical lower bound in terms of peak memory consumption, we should definitely recommend a hardware configuration, that will at least sustain the default configuration for regular usage.

3. The default docker compose configuration should work out of the box from a clean checkout on whatever machine configuration we recommend. It should also be accompanied with a guide, how to adapt for larger deployments (e.g. change certain limits and set it according to a formula).

I think 3 contradicts with your statement about the docker-compose,
> this file should not just be used off the shelf as-is, only people with docker/compose experience should use it, and should customize it to their needs

I think the file should be usable and working, ideally running on CI. That's the whole point of infrastructure as code. Once a user changes the config-docker.yml, they also need to adapt the docker-compose.yml and choose appropriate target hardware.

> Should be be in a separate compose with other tweaks for this use case? 

I understand your preference as a maintainer to need to adapt as little as possible from the default configuration to support your own use case. But out of a "new user" perspective it would make sense that the default configuration is consistent and coherent.

Happy to discuss more, maybe let's chat on slack in the evening tonight.

Cheers

Reply:
trying to make docker-compose work off the shelf for all use cases, seems like a crazy amount of work to setup and maintain. Also, your statement about a "default configuration" made me wonder what you think that is - there isn't one. I personally only collect trade data, and only for a handful of pairs, for a handful of exchanges. Others try 10+ exchanges with 20+ pairs for all data sources (trade, books, etc). Everyone has different use cases, so how can there be a recommended setup?

Reply:
@mschneider im usually around on slack, so feel free to message me to discuss further

Reply:
@mschneider just wondering if you were going to complete the change we discussed on slack? I can handle this and close this out if you like

Reply:
@mschneider just wanted to check on this again - I won't bother you endlessly, so just let me know what you'd like to do or I'll close this out